README
Tetris for Terminals - "tt" - Written by Mike Taylor
====================================================

(1)  Specification
==================

	This is a game written in C for Berkeley and System V UNIX
	machines.  It was written on a Sun4 and has been extensively
	played and tested on both this machine and a Sun3, both
	running SunOS 4.0.1, a Berkeley 4.3bsd-derived UNIX, but it
	should port with minimum difficulties to any Berkeley UNIX.
	It has also been ported to SCO Unix System V/386 3.2.2, and
	should therefore be portable to other System V UNIX systems.
	One potential difficulty to look out for is incompatibility of
	the high-score file between different architectures, (eg.
	sun3 and sun4). I've done all I can to make it work, but there
	are no guarantees, and you might be better compiling multiple-
	architecture "tt"s with different high-score tables.

(2)  Compilation
================

	The Makefile as included should be pretty much applicable
	anywhere.  Simply edit the file "tt.c" to give the pathname of
	the file you want to use for the high-score table
	(SCORE_FILE), and unless you are using the LOCKF #definition
	(in tt.h), the file you want to use as a lock for this table
	(LOCK_FILE, which should be in a publicly writeable
	directory).  If your machine is running System V or a similar
	UNIX, then you should make sure that CFLAGS (see the Makefile)
	contains the flag -DSYSV.

	Note that if you're on a System V Unix, you should add
	-DNO_NAP_SYSCALL to CFLAGS if you do not have the nap(S)
	system call.  Try looking for it in the -lx library.  If you
	don't have the usleep() system call on BSD Unix, then add
	-DNO_USLEEP_SYSCALL to CFLAGS.

	Now type "make".  The program will be compiled, producing a
	binary called "tt".  This can then be moved to a bin directory
	if required.

(3) Acknowledgements
====================

	The game Tetris was apparently designed by "a Russian
	Researcher".  Sadly, history, or at least the version of it
	that I heard, does not record his name, and it seems that the
	poor frog-head not only made no money from his brilliant idea,
	but also didn't get any fame.  'S'sad.	Still, I credit him
	with the original idea, whoever he may be.

	The inspiration for this program came from two recently-posted
	versions of Tetris, both of which run on Sun workstations
	only.  There are a few of these around here, but many people
	have no access to them.	 I wrote this version so that everyone
	could play.  It's the communist in me :-)

	The program design, planning (hah!), coding, and so on for
	this version was all done entirely by me, (Mike Taylor), and
	very smug I am about it, too.  The only exception to this is
	that the game-levels were accidentally co-invented by James
	"root@weed" Beckett, by pressing the space-bar too many times.

	Play-testing, which has been extensive even now, less than a
	week after I started writing, has been by many people, but
	special mention goes to Paul Metcalfe, Kenton Oatley, Harvey
	"Max" Thomson and Paul "Freddy" Capper.	 No mention *at all*,
	not even this one, goes to Mike "Sunny" Lessacher, who claims
	not to like Tetris.  ("Not like Tetris?	 What does that
	mean?" :-)

	All documentation is also by me.

Update: Thu Jul  5 16:04:21 BST 1990

	I've added a bug-fix supplied by Boris Lucre Goldowsky
	(boris@edu.rochester.psych.prodigal), so that the program now
	correctly chooses where to drop pieces at the beginning of
	games on high levels -- thanks, Boris!

Update: Thu May 16 15:16:15 PDT 1991

	Harvey Thompson (harveyt@santa-cruz.co.uk) really did all the
	work for System V portability (fingers crossed that nap(S) is
	actually a standard system call).

Update: Thu Jun  1 08:34:20 BST 2006 (Yes, fifteen years later)

	Ryan Jud Hughes (rjhughes@umich.edu) contributed code and
	documentationfor two nice new features: -g displays "ghost"
	pieces that show where the current piece would fall if
	dropped, and -N shows the next piece in the spare right-side
	area.

(4) The Legal Position
======================

	Look, I'm sorry, I *know* this bit is dull, but it has to be
	done, and it saves time and trouble for everyone if we just
	get on with it.	 here we go:

		The program "tt", its visual appearance, its code, its
		documentation, etc., are the intellectual property of
		Mike Taylor.  The program may be freely distributed,
		copied, modified, re-posted or whatever PROVIDED that
		the authorship and ownership remains clear, and that
		no-one makes any money from it without me knowing (and
		taking a hefty cut!)  Whoops, that's spoilt the
		official feel to it.  Never mind, you get the idea.
		Do what you want provided it isn't sneaky.

(5) The Game
============

	Tetris is one of those simple-but-compulsive games that you
	persistently find yourself wanting to play "just once more".
	I know, I know what you're thinking, "We've all heard that
	before!", but it just happens to be true on this occasion.
	The object of the game is simply to prevent a stack of blocks
	from building up to the top of the play-area, (to the left of
	the screen).  You do this by rotating and shifting the blocks
	as they fall, in such a way as to make them fit together as
	well as possible at the bottom.	 If you manage to complete a
	whole line, from right to left, then that row will disappear,
	and all the rows above will fall down into its place.  it is
	possible to get more than one row at once, and it is a
	wonderful feeling to get four at once!

	There are five types of block, (seven if you count mirror
	images), each of which is made up of four squares stuck
	together, (hence "Tetris", I assume).  These pieces score
	different numbers of points based on how difficult or
	otherwise it is to fit them into the place required.  The
	pieces are:

		  ######	"T-shape"		1 point.
		    ##

		<><><><>	"Long one"		2 points.

		()	{}
		()	{}	"L-shapes"		3 points.
		()()  {}{}

		  [][]		"Square"		4 points.
		  [][]

		%%	@@	"S-shapes"		5 points.
		%%%%  @@@@
		  %%  @@

	Once a piece has been positioned where you want it, it can be
	dropped into place immediately.	 There are no extra points for
	this, 'cos I think it would spoil the purity of the scoring
	system; it just speeds up the game.

	Hehehe, that reminds me, as the game progresses, it very very
	gradually speeds up, until, at scores of around 3000, it gets
	very difficult indeed ...  I mean *very* difficult!

	The keys that control the pieces, and other special keys
	(those used to Quit or Pause the game, or Refresh the screen)
	are shown on the screen.  They can be re-defined (except the
	Refresh key), as explained in the manual page.

(6)  Basic Hints
================

	Obviously, the aim is to keep the stack low, so at all times,
	you should be on the lookout for ways to fit the currently-
	falling piece onto those that have already fallen in such a
	way as to complete rows.  Removing rows is the key to success.

	However, if a lower section of the screen gets hopelessly
	hole-ridden, it is sometimes best to forget about it, and
	concentrate on building complete rows higher up, as this can
	often be the best way to make the holes lower down become
	available again.  A good player can take over from a game 2/3
	full of badly-packed pieces, and eventually wrestle it right
	back down to ground level again.

	Finally, there is a tendency among beginners to leave long,
	thin gaps down the sides of the screen, praying for a "Long
	one" to slide down it -- it is, of course, at precisely these
	times that a succession of S-shapes, all of the same
	handedness, comes pouring down the screen.  The moral is
	two-fold: (1) Beware of allowing such a situation to build up
	in the first place, it often isn't necessary, and (2), learn
	how to remove layers from higher up the screen with non-long
	pieces, so your screen doesn't fill up while you wait for that
	magical "Long one".

(7)  What to do if you don't like it
====================================

	Email me at the following address: mirk@uk.co.ssl
	In fact, email me if you *do* like it.	Email me with
	bouquets, brickbats, bugs, baboons, bachelors, blueberries,
	boathouses, er ... um, sorry, I seem to have got a bit carried
	away.  Anyway, the point is, let me know what you think of
	"tt", and it might just influence the next version.  Also, I
	am pretty good at replying to mail!

	That's it -- have fun!
______________________________________________________________________________
Mike Taylor - {Christ,M{athemat,us}ic}ian ...  Email to: mirk@uk.co.ssl
     "Imagine the universe perfect and whole and beautiful.  Then be sure
	of one thing: God has imagined it quite a bit better than you"


NEWS
$Id: NEWS,v 1.4 2009-06-18 15:22:33 mike Exp $

Release 15.1 (Thu Jun 18 16:23:14 BST 2009)
     --	Replace C++-style //-comments with C-style /*-comments.
     -- Replace false with 0.
     -- I think this is now once more good C.

Release 15.0 (Mon Jun  5 11:55:51 BST 2006)
     --	New "-g" option shows "ghost" pieces indicating where the
	current piece would fall if dropped.  Contributed by Ryan Jud
	Hughes <rjhughes@umich.edu>
     -- New "-N" option shows what the next piece will be.  Also
	contributed by Ryan Jud Hughes <rjhughes@umich.edu>

Release 14.1 (Tue Oct 18 18:14:56 BST 2005)
     --	The form() function in "form.c" (declared in "form.h") now
	uses ANSI-style <stdarg.h> instead of old-style <varargs.h>,
	as the latter is no longer supported on modern Linux systems
	such as Ubuntu "Hoary Hedgehog".
     --	"util.c" now includes <term.h> to avoid a compiler warning
	when tgetstr() is used undeclared.
     --	"screen.c" now includes <string.h> to avoid a compiler warning when
	strlen() is used undeclared.
     --	New material on the end of the "Testimonial" file.

Release 14.0 (Wed May 28 15:24:45 BST 2003)
     --	Pretty much as written in 1989.



screen.c
/***************************************************************************\
|*									   *|
|*  screen.c:	A version of Tetris to run on ordinary terminals,	   *|
|*		(ie., not needing a workstation, so should available	   *|
|*		to peasant Newwords+ users.  This module handles all	   *|
|*		the icky curses(3x) bits.				   *|
|*									   *|
|*  Author:	Mike Taylor (mirk@uk.co.ssl)				   *|
|*  Started:	Fri May 26 12:26:05 BST 1989				   *|
|*									   *|
\***************************************************************************/

#ifdef SYSV
#  undef SYSV	/* SysV curses actually defines SYSV */
#endif
#include <string.h>
#include <curses.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/time.h>

#include "screen.h"
#include "tt.h"
#include "pieces.h"
#include "utils.h"
#include "game.h"

char ghost_as[3] = "..";
char erase_as[3] = {BLANK_CHAR, BLANK_CHAR, 0};

/***************************************************************************\
|*									   *|
|*  The function myrefresh() calls the curses(3x) function refresh()	   *|
|*  after first moving the cursor out of harm's way at the bottom of	   *|
|*  the screen.								   *|
|*									   *|
\***************************************************************************/

void myrefresh ()
{
  int x;
  
  if ((x = screen_depth-2) < GAME_DEPTH+1)
    x = GAME_DEPTH+1;
  
  move (x, 0);
  refresh ();
}

/***************************************************************************\
|*									   *|
|*  The function hoopy_refresh() touches the screen, then refreshes it,	   *|
|*  so curses(3X) doesn't get any chance to phlegm about with only doing   *|
|*  the bits that it thinks are OK.  So this should fix up screens that	   *|
|*  have been interfered with by biff(1), mesg(1) etc.			   *|
|*									   *|
|*  Hmm, it seems that my original plan is insufficient to make curses	   *|
|*  get its act together, so I have to physically re-draw the screen.	   *|
|*  The function now also takes parameters describing any piece that is	   *|
|*  active, and if the piece-number is non-negative, then the piece is	   *|
|*  drawn (with draw_piece())						   *|
|*									   *|
\***************************************************************************/

void hoopy_refresh (piece_no, orient, x, y, next_piece_no, next_orient)
  int piece_no, orient, x, y, next_piece_no, next_orient;
{
  clear ();
  setup_screen ();
  draw_board ();
  update_scores ();
  if (piece_no > -1)
    draw_piece (piece_no, orient, x, y, PD_DRAW, 0);
  if (next_piece_no > -1 && SHOW_NEXT)
    draw_piece (next_piece_no, next_orient, NEXT_Y, NEXT_X, PD_DRAW, 0);
  myrefresh ();
}

/***************************************************************************\
|*									   *|
|*  The function print_msg() prints a short message centered on the	   *|
|*  floor of the playing area, with a space before and after it.	   *|
|*  If the message is NULL (ie. (char*)0), or null, (ie. ""), then	   *|
|*  no message is printed.						   *|
|*									   *|
\***************************************************************************/

void print_msg (line)
  char *line;
{
  int i;

  move (GAME_DEPTH, 2);
  for (i = 0; i < 2*GAME_WIDTH; i++)
    addch (FLOOR_CHAR);

  if ((line != NULL) && (*line != '\0'))
    mvaddstr (GAME_DEPTH, GAME_WIDTH+1-(strlen (line))/2,
	      form (" %s ", line));
  myrefresh ();
}

/***************************************************************************\
|*									   *|
|*  The function clear_area() fills the playing area with BLANK_CHARs.	   *|
|*  It is used to clear the screen before each game, and also clearing	   *|
|*  the screen while displaying pieces one at a time when in debug mode.   *|
|*									   *|
\***************************************************************************/

void clear_area ()
{
  int i, j, status, fastflag = 0;
  fd_set fds;
  struct timeval tmout;
  static char buffer[LINELEN];

  for (i = 0; i < GAME_DEPTH; i++) {
    move (i, 2);
    for (j = 0; j < 2*GAME_WIDTH; j++)
      addch (BLANK_CHAR);
    
    if (fastflag == 0) {
      move (i+1, 2);
      for (j = 0; j < 2*GAME_WIDTH; j++)
	addch (FLOOR_CHAR);
      myrefresh ();

      FD_ZERO (&fds);
      FD_SET (0, &fds);
      tmout.tv_sec = 0L;	/* Future implementations of select(2) */
      tmout.tv_usec = 50000L; /* might change this value on return */
      if ((status = select (1, &fds, (fd_set*) NULL, (fd_set*) NULL,
			    &tmout)) == -1)
	if (errno != EINTR)
	  die (LE_SELECT, "select(2) failed in clear_area()");
      
      if (status != 0) {
	fastflag = 1;
	(void) read (0, buffer, LINELEN);
      }
    }
  }
  
  move (GAME_DEPTH, 2);
  for (j = 0; j < 2*GAME_WIDTH; j++)
    addch (FLOOR_CHAR);
  
  mvaddch (GAME_DEPTH, 0, CORNER_CHAR);
  addch (CORNER_CHAR);
  mvaddch (GAME_DEPTH, 2*GAME_WIDTH+2, CORNER_CHAR);
  addch (CORNER_CHAR);
  
  myrefresh ();
}

/***************************************************************************\
|*									   *|
|*  The function setup_screen should be called exactly once, near the	   *|
|*  beginning of execution.  It initialises curses(3x), and prints the	   *|
|*  game titles, the walls and the floor of the game area, and clears	   *|
|*  this area using clear_area() (perhaps unsurprisingly)		   *|
|*									   *|
|*  STOP PRESS: It no longer calls clear_area(), since play_game()	   *|
|*	does that fairly immediately after this function returns.	   *|
|*									   *|
|*  STOP PRESS^2: It now does do it again, since play_game() does the	   *|
|*	sneaky-but-slow clear that it fine at times, but cruddy for	   *|
|*	initialisation.							   *|
|*									   *|
\***************************************************************************/

void setup_screen ()
{
  int i;

  for (i = 0; i < GAME_DEPTH; i++) {
    mvaddch (i, 0, WALL_CHAR);
    addch (WALL_CHAR);
    mvaddch (i, 2*GAME_WIDTH+2, WALL_CHAR);
    addch (WALL_CHAR);
  }

  move (GAME_DEPTH, 2);
  for (i = 0; i < 2*GAME_WIDTH; i++)
    addch (FLOOR_CHAR);
  
  mvaddch (GAME_DEPTH, 0, CORNER_CHAR);
  addch (CORNER_CHAR);
  mvaddch (GAME_DEPTH, 2*GAME_WIDTH+2, CORNER_CHAR);
  addch (CORNER_CHAR);
  
  /* XXX -- Harv's SYSV port uses standout()/standend() here */
  mvaddstr (0, 2*GAME_WIDTH+6, form ("%sTETRIS FOR TERMINALS%*s%s",
				     so_str, so_gunk, "", se_str));
  mvaddstr (2, 2*GAME_WIDTH+6, "Written by Mike Taylor");
  mvaddstr (3, 2*GAME_WIDTH+6, "Email: mirk@uk.co.ssl");
  mvaddstr (4, 2*GAME_WIDTH+6, "Started: Fri May 26 12:26:05 BST 1989");
  mvaddstr (6, 2*GAME_WIDTH+6, form ("Game level: %d  Game mode: %d",
				     game_level, game_mode));
  mvaddstr (8, 2*GAME_WIDTH+6, "Score:");
  mvaddstr (9, 2*GAME_WIDTH+6, "Pieces:");
  mvaddstr (10, 2*GAME_WIDTH+6, "Levels:");
  if (SHOW_NEXT)
    mvaddstr (NEXT_Y-1, 2*GAME_WIDTH+28, "Next:");

  mvaddstr (12, 2*GAME_WIDTH+8, "Use keys:");
  mvaddstr (13, 2*GAME_WIDTH+8, "=========");
  mvaddstr (14, 2*GAME_WIDTH+8, form ("Move left:  '%c'", left_key));
  mvaddstr (15, 2*GAME_WIDTH+8, form ("Move right: '%c'", right_key));
  mvaddstr (16, 2*GAME_WIDTH+8, form ("Rotate:     '%c'", rotate_key));
  mvaddstr (17, 2*GAME_WIDTH+8, form ("Drop:       '%c'", drop_key));
  mvaddstr (18, 2*GAME_WIDTH+8, form ("Pause:      '%c'", susp_key));
  mvaddstr (19, 2*GAME_WIDTH+8, form ("Quit:       '%c'", quit_key));
  mvaddstr (20, 2*GAME_WIDTH+8, "Refresh:    '^L'");
}

/***************************************************************************\
|*									   *|
|*  The function setup_curses should be called exactly once, near the	   *|
|*  beginning of execution.  It initialises curses(3x), and notes that	   *|
|*  this has been done, by setting the global variable in_curses.	   *|
|*									   *|
\***************************************************************************/

void setup_curses ()
{
  initscr ();
  clear ();
#ifndef lint
  noecho ();
  crmode ();
#endif /*lint*/
  in_curses = 1;
}

/***************************************************************************\
|*									   *|
|*  The function update_scores() puts the sepecified values of score,	   *|
|*  no_pieces and no_levels on the screen in the specified positions.	   *|
|*									   *|
\***************************************************************************/

void update_scores ()
{
  int i;
  move (8, 2*GAME_WIDTH + 14);
  for (i=0; i<SCORE_WIDTH; i++) addch (BLANK_CHAR);
  addstr (form ("%d", score));

  move (9, 2*GAME_WIDTH + 14);
  for (i=0; i<SCORE_WIDTH; i++) addch (BLANK_CHAR);
  addstr (form ("%d", no_pieces));

  move (10, 2*GAME_WIDTH + 14);
  for (i=0; i<SCORE_WIDTH; i++) addch (BLANK_CHAR);
  addstr (form ("%d", no_levels));
}

/***************************************************************************\
|*									   *|
|*  The function draw_board() puts the current state of the global array   *|
|*  board[] ontop the curses(3x) screen, then refresh()es it.		   *|
|*									   *|
\***************************************************************************/

void draw_board ()
{
  int i, j;

  for (i = 0; i < GAME_DEPTH; i++)
    for (j = 0; j < GAME_WIDTH; j++)
      if (board[i+4][j] == PI_EMPTY) {
	mvaddch (i, 2*j+2, BLANK_CHAR);
	addch (BLANK_CHAR);
      }
      else
	mvaddstr (i, 2*j+2, pieces[board[i+4][j]].app);
}

/***************************************************************************\
|*									   *|
|*  The function draw_raw draws one of the tetris pieces on the screen in  *|
|*  a specified orientation and position and string.  It can thus be used  *|
|*  for drawing, erasing, or drawing a ghost of a piece, by passing the    *|
|*  appropriate arg as "str".						   *|
|*  This function is also called by draw_piece to do all of these things.  *|
|*  The form of the function is:					   *|
|*									   *|
|*      draw_raw (piece_no, orient, y, x, str)				   *|
|*									   *|
|*  All the arrguments are integers.  Common values of str are erase_as    *|
|*  (for erasing), ghost_as (for drawing the ghost), or			   *|
|*  pieces[piece_no].app (for drawing the piece itself.  Piece_no is	   *|
|*  between 0 and 6 inclusive, and specifies what sort of piece is	   *|
|*  required.  Orientation is between 0 and 3 inclusive, and states	   *|
|*  which way up the piece is to be drawn, and y and x express the	   *|
|*  position as an index into the GAME_DEPTH by GAME_WIDTH array	   *|
|*  that is the board.							   *|
|*									   *|
\***************************************************************************/
void draw_raw (piece_no, orient, y, x, str)
  int piece_no;
  int orient;
  int y;
  int x;
  char *str;
{
  int i;
  for (i = 0; i < NO_SQUARES; i++) {
    if (y+pieces[piece_no].index[orient][i][0] >= 0) {
      mvaddstr (y+pieces[piece_no].index[orient][i][0],
	        (2*(x+pieces[piece_no].index[orient][i][1]))+2,
	        str);
    }
  }
}

/***************************************************************************\
|*									   *|
|*  draw_piece is the drawing function that should usually be called from  *|
|*  the outside.  It draws (or erases) the piece (and ghost), all	   *|
|*  regular-like.  It just makes the appropriate calls to draw_raw (after  *|
|*  figuring out where the ghost should be).				   *|
|*  The only situation to use draw_raw instead of draw_piece is if you     *|
|*  don't want the ghost to show up, such as for drawing the "Next"        *|
|*  piece.								   *|
|*  All the arrguments are integers.  Flag is either PD_DRAW or		   *|
|*  PD_ERASE, specifying the effect of the function.  The rest are as in   *|
|*  draw_raw.								   *|
|*									   *|
\***************************************************************************/
void draw_piece (piece_no, orient, y, x, flag)
  int piece_no;
  int orient;
  int y;
  int x;
  int flag;
{
  int j;
  
  if (GHOST_PIECES) 
    j = ghost_height(piece_no, orient, y, x);

  if (flag == PD_ERASE) {
    if (GHOST_PIECES)
      draw_raw (piece_no, orient, j, x, erase_as );
    draw_raw (piece_no, orient, y, x, erase_as );
  } /* if we're erasing */
  else {
    if (GHOST_PIECES) 
      draw_raw (piece_no, orient, j, x, ghost_as );
    draw_raw (piece_no, orient, y, x, pieces[piece_no].app );
  } /* else we are drawing */
}


/***************************************************************************\
|*									   *|
|*  The function place_piece takes the same parameters as draw_piece,	   *|
|*  except for the flag, and does not draw the piece, but places it	   *|
|*  on the board.  No checking is done to see if it will fit, since	   *|
|*  should already have been done by can_place().			   *|
|*									   *|
\***************************************************************************/

void place_piece (piece_no, orient, y, x)
  int piece_no;
  int orient;
  int y;
  int x;
{
  int i;

  for (i = 0; i < NO_SQUARES; i++)
    board[y+4+pieces[piece_no].index[orient][i][0]]
	 [x+pieces[piece_no].index[orient][i][1]] = piece_no;
}

/***************************************************************************\
|*									   *|
|*  The function can_place takes the same parameters as place_piece,	   *|
|*  It does not draw the piece, nor place it on the board, but returns	   *|
|*  an integer value -- 0 if the piece will not fit on the board, 1 if	   *|
|*  it will (with the specified orientation, position, etc.)		   *|
|*									   *|
\***************************************************************************/

#define HERE(x) pieces[piece_no].index[orient][i][x]

int can_place (piece_no, orient, y, x)
  int piece_no;
  int orient;
  int y;
  int x;
{
  int i;

  for (i = 0; i < NO_SQUARES; i++)
    if (((x+HERE(1) >= GAME_WIDTH) ||	/* Off right of screen or */
	 (x+HERE(1) < 0)) ||		/* Off left of screen */
	(y+HERE(0) >= GAME_DEPTH) ||	/* Off bottom of screen */
	(board[y+4+HERE(0)][x+HERE(1)] != PI_EMPTY))
					/* Board position not empty */
      return (0);
  
  return (1);
}


/***************************************************************************\
|*									   *|
|*  The function ghost_height takes piece_no, orient, y, and x, all	   *|
|*  integers describing the type, orientation, and location of a piece),   *|
|*  and returns the height at which we should draw the "ghost" piece.	   *|
|*  The ghost piece is the one that shows where it would be if we dropped  *|
|*  the piece right this instant.					   *|
|*									   *|
\***************************************************************************/
int ghost_height(int piece_no, int orient, int y, int x) 
{
    int j;
    for(j = y; j < GAME_DEPTH; j++) {
      if( !can_place(piece_no, orient, j, x) ) { break; }
    } /* for from here to the bottom of the board */

    j--;
    return j;
}


/*-------------------------------------------------------------------------*/
